### PR TITLE
Blazegraph views: incoming/outgoing routes

### LINKED ISSUE #1868: Blazegraph Plugin: define the plugin routes

### PR DESCRIPTION
**Pull request by wwajerowicz**
Fixes #1868

### CODE REVIEW

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/BlazegraphViewsQuery.scala [hunk 1]
```diff
@@ -112,10 +213,11 @@ object BlazegraphViewsQuery {
...
-  private[blazegraph] type FetchView = (IdSegment, ProjectRef) => IO[BlazegraphViewRejection, ViewResource]
+  private[blazegraph] type FetchView    = (IdSegment, ProjectRef) => IO[BlazegraphViewRejection, ViewResource]
+  private[blazegraph] type FetchProject = ProjectRef => IO[BlazegraphViewRejection, Project]
 
-  final def apply(acls: Acls, views: BlazegraphViews, client: BlazegraphClient)(implicit
+  final def apply(acls: Acls, views: BlazegraphViews, projects: Projects, client: BlazegraphClient)(implicit
       config: ExternalIndexingConfig
   ): BlazegraphViewsQuery =
-    new BlazegraphViewsQueryImpl(views.fetch, acls, client)
+    new BlazegraphViewsQueryImpl(views.fetch, pRef => projects.fetchProject(pRef), acls, client)
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/BlazegraphViewsQuery.scala:222]: `projects.fetchProject` should work
or `projects.fetchProject[ProjectNotFound]`

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/model/SparqlLink.scala [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    * @param types         the collection of types of this resource
+    * @param deprecated    whether the resource is deprecated of not
+    * @param created       the instant when this resource was created
+    * @param updated       the last instant when this resource was updated
+    * @param createdBy     the identity that created this resource
+    * @param updatedBy     the last identity that updated this resource
+    * @param constrainedBy the schema that this resource conforms to
+    * @param paths         the paths from where the link has been found
+    */
+  final case class SparqlResourceLink(
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/model/SparqlLink.scala:48]: All of this is very suspiciously similar to `ResourceF`. Couldn't we reuse it and have a `ResourceF[A]` where A in your case is `paths` ?

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/model/SparqlLink.scala [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      val types = bindings.get("types").map(binding => toIris(binding.value).toSet).getOrElse(Set.empty)
+      val paths = bindings.get("paths").map(binding => toIris(binding.value).toList).getOrElse(List.empty)
+      bindings.get("s").map(_.value).flatMap(Iri.absolute(_).toOption).map(SparqlExternalLink(_, paths, types))
+    }
+  }
+
+  private def toIris(string: String): Array[Iri] =
+    string.split(" ").flatMap(Iri.absolute(_).toOption)
+
+  implicit val linkEncoder: Encoder.AsObject[SparqlLink] = Encoder.AsObject.instance {
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/model/SparqlLink.scala:115]: this a duplicated from the `ResourceF`

**Review by umbreak — CHANGES_REQUESTED** [review]

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/BlazegraphViewsQuery.scala [hunk 4]
```diff
@@ -2,24 +2,67 @@ package ch.epfl.bluebrain.nexus.delta.plugins.blazegraph
...
+    *
+    * @param id         the resource identifier
+    * @param projectRef the project of the resource
+    * @param pagination the pagination config
+    */
+  def incoming(
+      id: IdSegment,
+      projectRef: ProjectRef,
+      pagination: FromPagination
+  )(implicit caller: Caller, scheduler: Scheduler): IO[BlazegraphViewRejection, SearchResults[SparqlLink]]
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/BlazegraphViewsQuery.scala:49]: This should return a `SearchResults[ResourceF[A]]`

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/BlazegraphViewsQuery.scala [hunk 5]
```diff
@@ -2,24 +2,67 @@ package ch.epfl.bluebrain.nexus.delta.plugins.blazegraph
...
+    * @param projectRef           the project of the resource
+    * @param pagination           the pagination config
+    * @param includeExternalLinks whether to include links to resources not managed by Delta
+    */
+  def outgoing(
+      id: IdSegment,
+      projectRef: ProjectRef,
+      pagination: FromPagination,
+      includeExternalLinks: Boolean
+  )(implicit caller: Caller): IO[BlazegraphViewRejection, SearchResults[SparqlLink]]
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/BlazegraphViewsQuery.scala:64]: same

**Review by umbreak — CHANGES_REQUESTED** [review]

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/model/SparqlLink.scala (hunk 2 above)
Comment (wwajerowicz) [reply delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/model/SparqlLink.scala:48]: We have two cases: 
`SparqlResourceLink` and`SparqlExternalLink`. `SparqlResourceLink` is almost ResourceF and probably we could use it for that, but SparqlExternalLink has only `id`, `paths` and `types`  and in this case ResourceF contains a lot of fields which are not available in `SparqlExternalLink`.

Comment (umbreak) [reply delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/model/SparqlLink.scala:48]: Yeah. So the case `SparqlResourceLink` should wrap a resourceF

**Review by umbreak — CHANGES_REQUESTED** [review]

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/model/SparqlLink.scala (hunk 2 above)
Comment (wwajerowicz) [reply delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/model/SparqlLink.scala:48]: What would be the value of `rev`, `deprecated`, `createdAt`, `updatedAt`, `createdBy`, `updatedBy` and `schema` for a `ResourceF` which would represent what we currently have as `SparqlExternalLink`?

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/BlazegraphViewsQuery.scala (hunk 4 above)
Comment (umbreak) [reply delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/BlazegraphViewsQuery.scala:49]: Ok I see why it cannot return a ResourceF. Ignore this, but make `final case class SparqlResourceLink(resource: ResourceF[A])`

File: delta/plugins/blazegraph/src/main/resources/blazegraph/incoming.txt [hunk 6]
```diff
@@ -1,13 +1,14 @@
...
 
 WITH {
-  SELECT ?s ((GROUP_CONCAT(?p)) as ?paths) ?types ?_self ?_constrainedBy ?_project ?_rev ?_deprecated ?_createdAt  ?_createdBy ?_updatedAt ?_updatedBy {
-    SELECT ?s ?p (GROUP_CONCAT(DISTINCT ?type) AS ?types) ?_self ?_constrainedBy ?_project ?_rev ?_deprecated ?_createdAt  ?_createdBy ?_updatedAt ?_updatedBy (count(?mid) as ?pos) WHERE {
+  SELECT ?s ((GROUP_CONCAT(?p)) as ?paths) ?types ?_self ?_constrainedBy ?_schemaProject ?_project ?_rev ?_deprecated ?_createdAt  ?_createdBy ?_updatedAt ?_updatedBy {
+    SELECT ?s ?p (GROUP_CONCAT(DISTINCT ?type) AS ?types) ?_self ?_constrainedBy ?_schemaProject ?_project ?_rev ?_deprecated ?_createdAt  ?_createdBy ?_updatedAt ?_updatedBy (count(?mid) as ?pos) WHERE {
       GRAPH ?g1 {
         ?s nxv:self ?_self .
         ?s nxv:constrainedBy ?_constrainedBy .
+        ?s nxv:schemaProject ?_schemaProject .
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/resources/blazegraph/incoming.txt:11]: This field must be optional (only resources have `schemaProject`, views, resolvers, etc don't).

File: delta/plugins/blazegraph/src/main/resources/blazegraph/outgoing_scoped.txt [hunk 7]
```diff
@@ -14,6 +14,7 @@ WITH {
       GRAPH ?g2 {
         ?s nxv:self ?_self .
         ?s nxv:constrainedBy ?_constrainedBy .
+        ?s nxv:schemaProject ?_schemaProject .
```
Comment (umbreak) [inline delta/plugins/blazegraph/src/main/resources/blazegraph/outgoing_scoped.txt:17]: same

**Review by umbreak — CHANGES_REQUESTED** [review]

**Review by umbreak — APPROVED** [review]

**Pull request merged**